rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,321 @@
+            // now that we have full chains in hand.
+            if matches!(role, Role::EndEntity | Role::Client) {
+                let root = &certified_keys
+                    .get(&(Role::TrustAnchor, alg.inner))
+                    .unwrap()
+                    .cert;
+                let intermediate = &certified_keys
+                    .get(&(Role::Intermediate, alg.inner))
+                    .unwrap()
+                    .cert;
+
+                // Write the PEM chain and full chain files for the end entity and client certs.
+                // Chain files include the intermediate and root certs, while full chain files include
+                // the end entity or client cert as well.
+                for f in [
+                    ("chain", &[intermediate, root][..]),
+                    ("fullchain", &[&cert, intermediate, root][..]),
+                ] {
+                    let mut chain_file = File::create(alg.output_directory().join(format!(
+                        "{}.{}",
+                        role.label(),
+                        f.0
+                    )))?;
+                    for cert in f.1 {
+                        chain_file.write_all(cert.pem().as_bytes())?;
+                    }
+                }
+            }
+
+            certified_keys.insert((role, alg.inner), CertifiedKey { cert, key_pair });
+        }
+    }
+
+    Ok(())
+}
+
+fn crl_for_serial(serial_number: &SerialNumber) -> CertificateRevocationListParams {
+    let now = OffsetDateTime::now_utc();
+    CertificateRevocationListParams {
+        this_update: now,
+        next_update: now + Duration::from_secs(60 * 60 * 24 * 5),
+        crl_number: SerialNumber::from(1234),
+        issuing_distribution_point: None,
+        revoked_certs: vec![RevokedCertParams {
+            serial_number: serial_number.clone(),
+            revocation_time: now,
+            reason_code: Some(RevocationReason::KeyCompromise),
+            invalidity_date: None,
+        }],
+        key_identifier_method: KeyIdMethod::Sha256,
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum Role {
+    Client,
+    EndEntity,
+    Intermediate,
+    TrustAnchor,
+}
+
+impl Role {
+    fn key_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.key", self.label()))
+    }
+
+    fn cert_pem_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.cert", self.label()))
+    }
+
+    fn cert_der_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.der", self.label()))
+    }
+
+    fn label(&self) -> &'static str {
+        match self {
+            Self::Client => "client",
+            Self::EndEntity => "end",
+            Self::Intermediate => "inter",
+            Self::TrustAnchor => "ca",
+        }
+    }
```

Comment:
Nit: move all these simple getters below `params()` and `common_name()`?

Reply:
Done.